Fast standalone symbolize for unwinding

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #123966

We've had issues using addr2line. On certain versions of
CentOS it is on a version that has a performance regression making it very slow,
and even normallly it is not that fast, taking several seconds even when parallelized
for a typical memory trace dump.

Folly Symbolize or LLVMSymbolize are fast but it requires PyTorch take a dependency on those libraries to do this, and given the number of environments we run stuff in, we end up hitting cases where we fallback to slow addr2line behavior.

This adds a standalone symbolizer to PyTorch similar to the unwinder which has
no external dependencies and is ~20x faster than addr2line for unwinding PyTorch frames.

I've tested this on some memory profiling runs using all combinations of {gcc, clang} x {dwarf4, dwarf5} and it seems to do a good job at getting line numbers and function names right. It is also careful to route all reads of library data through the `CheckedLexer` object, which ensure it is not reading out of bounds of the section. Errors are routed through UnwindError so that those exceptions get caught and we produce a ?? frame rather than crash. I also added a fuzz test which gives all our symbolizer options random addresses in the process to make sure they do not crash.

Differential Revision: [D56828968](https://our.internmc.facebook.com/intern/diff/D56828968)

zdevito committed:
[WIP] Fast standalone symbolize for unwinding

We've had issues using addr2line. On certain versions of
CentOS it is on a version that has a performance regression making it very slow,
and even normallly it is not that fast, taking several seconds even when parallelized
for a typical memory trace dump.

LLVMSymbolize is fast but it requires PyTorch take an LLVM dependency
which we have been trying to remove.

This adds a standalone symbolizer to PyTorch similar to the unwinder which has
not external dependencies and is ~20x faster than addr2line for unwinding PyTorch frames.

[ghstack-poisoned]

zdevito committed:
Update on "[WIP] Fast standalone symbolize for unwinding"


We've had issues using addr2line. On certain versions of
CentOS it is on a version that has a performance regression making it very slow,
and even normallly it is not that fast, taking several seconds even when parallelized
for a typical memory trace dump.

LLVMSymbolize is fast but it requires PyTorch take an LLVM dependency
which we have been trying to remove.

This adds a standalone symbolizer to PyTorch similar to the unwinder which has
not external dependencies and is ~20x faster than addr2line for unwinding PyTorch frames.

[ghstack-poisoned]

zdevito committed:
Update on "[WIP] Fast standalone symbolize for unwinding"


We've had issues using addr2line. On certain versions of
CentOS it is on a version that has a performance regression making it very slow,
and even normallly it is not that fast, taking several seconds even when parallelized
for a typical memory trace dump.

LLVMSymbolize is fast but it requires PyTorch take an LLVM dependency
which we have been trying to remove.

This adds a standalone symbolizer to PyTorch similar to the unwinder which has
not external dependencies and is ~20x faster than addr2line for unwinding PyTorch frames.

[ghstack-poisoned]

zdevito committed:
Update on "Fast standalone symbolize for unwinding"


We've had issues using addr2line. On certain versions of
CentOS it is on a version that has a performance regression making it very slow,
and even normallly it is not that fast, taking several seconds even when parallelized
for a typical memory trace dump.

LLVMSymbolize is fast but it requires PyTorch take an LLVM dependency
which we have been trying to remove.

This adds a standalone symbolizer to PyTorch similar to the unwinder which has
not external dependencies and is ~20x faster than addr2line for unwinding PyTorch frames.

[ghstack-poisoned]

wconstab commented on test/profiler/test_profiler.py:
```diff
@@ -3561,6 +3566,70 @@ def test_profiler_pattern_matcher_json_report(self):
+                        (section_size,) = unpack(
+                            "Q", section_headers_start + i * section_header_size + 32
+                        )
+                        start = int(filename.split("-")[0], 16) + section_offset
```
what is the filename representing- does each mapped library have a filename representing its start address in memory?

zdevito replied:
A file looks like this:
`lr-------- 1 zdevito users 64 Apr 22 09:11 7fcea5800000-7fcea5828000 -> /usr/lib64/libc.so.6`

wconstab commented on test/profiler/test_profiler.py:
```diff
@@ -3561,6 +3566,70 @@ def test_profiler_pattern_matcher_json_report(self):
+        r.seed(1)
+        text_sections = get_text_sections()
+        addrs = []
+        for i in range(200):
```
ok so basically, take any random byte of code and ask for its unwind, make sure this doesn't crash.  seems reasonable

wconstab commented on torch/csrc/profiler/unwind/dwarf_symbolize_enums.h:
```diff
@@ -0,0 +1,181 @@
+#include <optional>
+
+enum {
+  DW_TAG_subprogram = 0x2e,
```
isn't there some dwarf header we can include that defines all these things? i guess we don't have access to it though?

zdevito replied:
It doesn't come by default. However the numbers themselves are defined by the standard, so they can't change anyway.

wconstab commented on torch/csrc/utils/cpp_stacktraces.cpp:
```diff
@@ -47,9 +47,31 @@ bool get_cpp_stacktraces_enabled() {
-  static bool disabled = compute_disable_addr2line();
-  return disabled;
+static torch::unwind::Mode compute_symbolize_mode() {
+  auto envar_c = std::getenv("TORCH_SYMBOLIZE_MODE");
```
we already ensure this env is set to something? or where is the default provided

wconstab commented on torch/csrc/utils/cpp_stacktraces.cpp:
```diff
@@ -47,9 +47,31 @@ bool get_cpp_stacktraces_enabled() {
+          envar);
+    }
+  } else {
+    return compute_disable_addr2line() ? unwind::Mode::dladdr
```
i assume this is the status quo today, we default to dladdr only if we set the env for addr2line disablement

zdevito replied:
yeah, and once this lands we can delete DISABLE_ADDR2LINE in favor of just setting the mode. Hopefully we can enable the fast version by default and get rid of the flags.

ezyang commented:
Not sure how you want to run review here, an actual careful review will take some time since I don't know how DWARF works. Shall we just dropship the code in?

zdevito commented:
For review, I think it is fine if the review just makes sure this doesn't break anything when disabled, and we can land it an experiment with it on some real runs by setting environment flags. No need to go deeply into DWARF. If you have time, a quick check looking for "this can't crash if the dwarf information is bogus" would be good. I tried to put all accesses through the `CheckedLexer` which can't read out of bounds of a segment.

zdevito committed:
Update on "Fast standalone symbolize for unwinding"


We've had issues using addr2line. On certain versions of
CentOS it is on a version that has a performance regression making it very slow,
and even normallly it is not that fast, taking several seconds even when parallelized
for a typical memory trace dump.

Folly Symbolize or LLVMSymbolize are fast but it requires PyTorch take a dependency on those libraries to do this, and given the number of environments we run stuff in, we end up hitting cases where we fallback to slow addr2line behavior.

This adds a standalone symbolizer to PyTorch similar to the unwinder which has
no external dependencies and is ~20x faster than addr2line for unwinding PyTorch frames.

I've tested this on some memory profiling runs using all combinations of {gcc, clang} x {dwarf4, dwarf5} and it seems to do a good job at getting line numbers and function names right. It is also careful to route all reads of library data through the `CheckedLexer` object, which ensure it is not reading out of bounds of the section. Errors are routed through UnwindError so that those exceptions get caught and we produce a ?? frame rather than crash. I also added a fuzz test which gives all our symbolizer options random addresses in the process to make sure they do not crash.

[ghstack-poisoned]

ezyang commented on torch/csrc/profiler/unwind/unwind_error.h:
```diff
@@ -1,6 +1,31 @@
+#define PRINT_INST(...)
+
+// #define PRINT_LINE_TABLE(...) LOG_INFO(__VA_ARGS__)
+#define PRINT_LINE_TABLE(...)
```
would be nice if these could go into our glog logging permanently as debug but whatever lol

ezyang commented on torch/csrc/profiler/python/init.cpp:
```diff
@@ -606,6 +603,33 @@ void initPythonBindings(PyObject* module) {
+        } else if (mode_s == "addr2line") {
+          mode = torch::unwind::Mode::addr2line;
+        } else if (mode_s == "dladdr") {
+          mode = torch::unwind::Mode::dladdr;
```
a bit dupey with the envvar parsing code

ezyang reviewed (approved):
Integration code looks fine. I didn't review contents of unwind/ dir

zdevito commented:
@pytorchbot merge

zdevito commented:
@pytorchbot help

zdevito commented:
@pytorchbot --help

zdevito commented:
@pytorchbot merge -r

zdevito commented:
@pytorchbot merge

zdevito commented:
@pytorchbot merge -i

jeanschmidt commented:
@pytorchbot revert -m "Breaking internal builds, check D56522678" -c nosignal

zdevito commented:
@pytorchbot rebase

zdevito committed:
Update on "Fast standalone symbolize for unwinding"


We've had issues using addr2line. On certain versions of
CentOS it is on a version that has a performance regression making it very slow,
and even normallly it is not that fast, taking several seconds even when parallelized
for a typical memory trace dump.

Folly Symbolize or LLVMSymbolize are fast but it requires PyTorch take a dependency on those libraries to do this, and given the number of environments we run stuff in, we end up hitting cases where we fallback to slow addr2line behavior.

This adds a standalone symbolizer to PyTorch similar to the unwinder which has
no external dependencies and is ~20x faster than addr2line for unwinding PyTorch frames.

I've tested this on some memory profiling runs using all combinations of {gcc, clang} x {dwarf4, dwarf5} and it seems to do a good job at getting line numbers and function names right. It is also careful to route all reads of library data through the `CheckedLexer` object, which ensure it is not reading out of bounds of the section. Errors are routed through UnwindError so that those exceptions get caught and we produce a ?? frame rather than crash. I also added a fuzz test which gives all our symbolizer options random addresses in the process to make sure they do not crash.

[ghstack-poisoned]

zdevito committed:
Update on "Fast standalone symbolize for unwinding"


We've had issues using addr2line. On certain versions of
CentOS it is on a version that has a performance regression making it very slow,
and even normallly it is not that fast, taking several seconds even when parallelized
for a typical memory trace dump.

Folly Symbolize or LLVMSymbolize are fast but it requires PyTorch take a dependency on those libraries to do this, and given the number of environments we run stuff in, we end up hitting cases where we fallback to slow addr2line behavior.

This adds a standalone symbolizer to PyTorch similar to the unwinder which has
no external dependencies and is ~20x faster than addr2line for unwinding PyTorch frames.

I've tested this on some memory profiling runs using all combinations of {gcc, clang} x {dwarf4, dwarf5} and it seems to do a good job at getting line numbers and function names right. It is also careful to route all reads of library data through the `CheckedLexer` object, which ensure it is not reading out of bounds of the section. Errors are routed through UnwindError so that those exceptions get caught and we produce a ?? frame rather than crash. I also added a fuzz test which gives all our symbolizer options random addresses in the process to make sure they do not crash.

[ghstack-poisoned]

zdevito committed:
Update on "Fast standalone symbolize for unwinding"


We've had issues using addr2line. On certain versions of
CentOS it is on a version that has a performance regression making it very slow,
and even normallly it is not that fast, taking several seconds even when parallelized
for a typical memory trace dump.

Folly Symbolize or LLVMSymbolize are fast but it requires PyTorch take a dependency on those libraries to do this, and given the number of environments we run stuff in, we end up hitting cases where we fallback to slow addr2line behavior.

This adds a standalone symbolizer to PyTorch similar to the unwinder which has
no external dependencies and is ~20x faster than addr2line for unwinding PyTorch frames.

I've tested this on some memory profiling runs using all combinations of {gcc, clang} x {dwarf4, dwarf5} and it seems to do a good job at getting line numbers and function names right. It is also careful to route all reads of library data through the `CheckedLexer` object, which ensure it is not reading out of bounds of the section. Errors are routed through UnwindError so that those exceptions get caught and we produce a ?? frame rather than crash. I also added a fuzz test which gives all our symbolizer options random addresses in the process to make sure they do not crash.

[ghstack-poisoned]

zdevito committed:
Update on "Fast standalone symbolize for unwinding"


We've had issues using addr2line. On certain versions of
CentOS it is on a version that has a performance regression making it very slow,
and even normallly it is not that fast, taking several seconds even when parallelized
for a typical memory trace dump.

Folly Symbolize or LLVMSymbolize are fast but it requires PyTorch take a dependency on those libraries to do this, and given the number of environments we run stuff in, we end up hitting cases where we fallback to slow addr2line behavior.

This adds a standalone symbolizer to PyTorch similar to the unwinder which has
no external dependencies and is ~20x faster than addr2line for unwinding PyTorch frames.

I've tested this on some memory profiling runs using all combinations of {gcc, clang} x {dwarf4, dwarf5} and it seems to do a good job at getting line numbers and function names right. It is also careful to route all reads of library data through the `CheckedLexer` object, which ensure it is not reading out of bounds of the section. Errors are routed through UnwindError so that those exceptions get caught and we produce a ?? frame rather than crash. I also added a fuzz test which gives all our symbolizer options random addresses in the process to make sure they do not crash.

[ghstack-poisoned]

zdevito commented:
@zdevito has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D56828968).

zdevito commented:
@pytorchbot merge
